source/language/openpulse.rst

```diff
@@ -0,0 +1,740 @@
+an array of complex samples or a parametric function supported directly by the hardware
+such as a Gaussian waveform.
+
+Arrays initialization is done as in OpenQASM.
+
+.. code-block:: c
+
+    envelope env = [0.1, 1.+1I, 0.1 - 0.2I, ...];
+
+Parametric pulses are initialized by assigning the result of``envelope``'s to the result of a (kernel) function call.
+
+.. code-block:: c
+
+  envelope env = gaussian(1.+1I, 1024dt, 128dt);
+
+The following parametric functions are currently defined
+
+.. code-block:: c
+
+    kernel gaussian(complex[size] amp, length duration, length sigma) -> envelope;
+    kernel gaussian_square(complex[size] amp, length duration, length sigma, length square_width) -> envelope;
+    kernel drag(complex[size] amp, length duration, length sigma, float[size] beta) -> envelope;
+    kernel constant(complex[size] amp, length duration) -> envelope;
+
+Carriers
+~~~~~~~~
+
+Carriers represent stateful modulation signals in hardware such as those used to implement the <Virtual Z-gate `https://arxiv.org/abs/1612.00858`>_.
+The standard carrier has the form :math:`Ae^{i(2*\pi*\imag*freq(t_i)*\Delta t +\phi(t_i) + \phi(t_{i-1})}`.
+:math:`A` is the amplitude. :math:`freq(t_i)` and :math:`\phi(t_i)` are the carrier's instantaneous frequency and
+relative phase respectively, and :math:`phi(t_{i-1})` is the carrier's accumulated phase which stores the history of the carrier.
+Carriers are declared by calling an exponential kernel function
+``kernel exp(complex[size] amp, float[size] freq, angle[size] phase) -> carrier``.
+
+.. code-block:: c
+
+    carrier carr = exp(0.5+0.5I, 5e9, pi);
+
+The instantaneous carrier parameters may be accessed via ``carrier.amp``, ``carrier.freq`` and ``carrier.phase``.
+Standard OpenQASM assignment and arithmetic operations apply to these fields--for instance,
+to increment the phase one may apply ``carrier.phase += pi;``. However, in contrast to the circuit model,
+care must be taken to the global time :math:`t_i` at which this phase increment occurs, as a carrier may
+be shared across many channels operating in parallel in time. Updates to carriers take place at the
+greatest time of all channels within the current scope. This enables reasoning about the absolute accumulated
+phase of the carrier and how this impacts signals that are emitted on different channels with a shared carrier.
+See the :ref:`Timing` for more detail on this behavior.
+
+It is the responsibility of the compiler to identify operations which affect the carrier state, and the
+relative timing at which the update takes place with respect to the target channel.
+
+Signal composition & transformation operations
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Complex signals may be constructed through signal transformation and composition operations applied to
+``envelope`` and ``carrier`` signals which are the signal production base types. This is very natural
```

Comment:
according to line 50, they can also be constructed directly?

Reply:
That is true--will make this clearer in frame proposal.